src/main/java/controllers/BookstoreRestController.java

```diff
@@ -0,0 +1,202 @@
+        Customer customer = new Customer(name, address, email, phoneNumber);
+        customerRepository.save(customer);
+        return customer;
+    }
+
+
+    //Shopping Cart REST endpoints
+    @GetMapping("/api/getShoppingCart")
+    public ShoppingCart getShoppingCart(@RequestParam(value = "id") long id) {
+        return shoppingCartRepository.findById(id);
+    }
+
+    @GetMapping("/api/getShoppingCarts")
+    public Iterable<ShoppingCart> getShoppingCarts() {
+        return shoppingCartRepository.findAll();
+    }
+
+    @GetMapping("/api/getShoppingCartsByCustomer")
+    public Iterable<ShoppingCart> getShoppingCartsByCustomer(@RequestParam(value = "customerID") long customerID) {
+        Customer customer = customerRepository.findById(customerID);
+        if (customer == null)
+            return null;
+        return shoppingCartRepository.findByCustomer(customer);
+    }
+
+    @PostMapping("/api/newShoppingCart")
+    public ShoppingCart newShoppingCart(@RequestParam(value = "customerID") long customerID) {
+        Customer customer = customerRepository.findById(customerID);
+        if (customer == null)
+            return null;
+        ShoppingCart shoppingCart = new ShoppingCart(customer);
+        shoppingCartRepository.save(shoppingCart);
+        customer.setShoppingCart(shoppingCart);
+        customerRepository.save(customer);
+        return shoppingCart;
+    }
+
+    @PostMapping("/api/addBookToShoppingCart")
+    public ShoppingCart addBookToShoppingCart(@RequestParam(value = "shoppingCartID") long shoppingCartID, @RequestParam(value = "bookID") long bookID) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findById(shoppingCartID);
+        if (shoppingCart == null)
+            return null;
+        Book book = bookRepository.findById(bookID);
+        if (book == null)
+            return null;
+        shoppingCart.addBook(book);
+        book.addShoppingCart(shoppingCart);
+        shoppingCartRepository.save(shoppingCart);
+        bookRepository.save(book);
+        return shoppingCart;
+    }
+
+
+    //Sale REST endpoints
+    @GetMapping("/api/getSale")
+    public Sale getSale(@RequestParam(value = "id") long id) {
+        return saleRepository.findById(id);
+    }
+
+    @GetMapping("/api/getSales")
+    public Iterable<Sale> getSales() {
+        return saleRepository.findAll();
+    }
+
+    @GetMapping("/api/getSalesByCustomer")
+    public Iterable<Sale> getSalesByCustomer(@RequestParam(value = "customerID") long customerID) {
+        Customer customer = customerRepository.findById(customerID);
+        if (customer == null)
+            return null;
+        return saleRepository.findByCustomer(customer);
+    }
+
+    @PostMapping("/api/newSale")
+    public Sale newSale(@RequestParam(value = "shoppingCartID") long shoppingCartID) {
```

Comment:
Same comment as previously about using a Customer to access the shopping cart.

Reply:
Addressed in 36c15e0107c2adc121290e2739d3baf67ad1b6d2